openam-core-rest/src/main/java/org/forgerock/openam/core/rest/IdentityResourceV4.java

```diff
@@ -0,0 +1,497 @@
+    }
+
+    /**
+     * Return every delegation privilege name configured in the supplied realm.
+     *
+     * @param realm the realm in path format
+     * @return An immutable set of privilege names
+     * @throws SSOException        if the admin SSO token cannot be obtained or is invalid
+     * @throws DelegationException If the delegation framework fails to read the privileges
+     */
+    private Set<String> getPrivilegeNames(String realm) throws SSOException, DelegationException {
+        SSOToken adminSSOToken = AccessController.doPrivileged(AdminTokenAction.getInstance());
+        DelegationManager mgr = new DelegationManager(adminSSOToken, realm);
+        return mgr.getConfiguredPrivilegeNames();
+    }
+
+    /**
+     * Extract the first universalid attribute value from an {@link IdentityDetails} instance.
+     */
+    private String getUniversalId(IdentityDetails identityDetails) {
+        return Arrays.stream(identityDetails.getAttributes())
+                .filter(attribute -> "universalid".equals(attribute.getName()))
+                .map(Attribute::getValues)
+                .filter(values -> values != null && values.length > 0)
+                .map(values -> values[0])
+                .findFirst()
+                .orElse(null);
+    }
+
+    /**
+     * Get the current state of a group from the IdM layer.
+     */
+    private IdentityDetails getExistingGroupDetails(Context context, String resourceId, String realm)
+            throws IdServicesException, SSOException {
+        return identityServices.read(resourceId,
+                IdentityRestUtils.getIdentityServicesAttributes(realm,GROUP_TYPE),
+                IdentityRestUtils.getSSOToken(RestUtils.getCookieFromServerContext(context)));
+    }
+
+    /**
+     * Get an {@link IdentityDetails} object that represents the desired post‑update state of a group, based on the
+     * supplied {@link UpdateRequest}. Only attributes relevant to membership changes are modified here; other
+     * attributes are copied verbatim from the existing group details.
+     */
+    private IdentityDetails getUpdatedGroupDetails(String resourceId, UpdateRequest request,
+            IdentityDetails groupDetails) throws BadRequestException {
+        IdentityDetails newGroupDetails = new IdentityDetails();
+        List<String> newMemberList = new ArrayList<>();
+        JsonValue memberList = request.getContent().get(new JsonPointer("members/uniqueMember"));
+        if (!memberList.isNull() && memberList.isList()) {
+            newMemberList = memberList.asList(String.class);
+        }
+        if (newMemberList.isEmpty()) {
+            Map<String, Set<String>> membersMap = new HashMap<>();
+            membersMap.put("uniqueMember", new HashSet<>());
+            newGroupDetails.setAttributes(IdentityServicesImpl.asAttributeArray(membersMap));
+        }
+        newGroupDetails.setMemberList(new ListWrapper(newMemberList.toArray(new String[0])));
```

Comment:
Please double check this update of group members. Do all UPDATE requests have to contain it? Because group members can be potentionaly in tens of thousands entries (I would consider extremes with e.g. million values).